### DISCUSSION TITLE
Testing zkSync Era rollup to a brand new EVM-compatible L1 chain (not Geth client)

### DISCUSSION THREAD

**Discussion by AlexRamRam** [zkSync-Community-Hub/zksync-developers#135] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

I’m trying to experiment with zkSync Era by instantiating a local zkSync L2 node configured to a local L1 test node of a new EVM-compatible chain (not a Geth client).

The documented [in-memory test node](https://era.zksync.io/docs/tools/testing/era-test-node.html) only supports forking from networks. So, I experimented with the [Dockerized setup](https://era.zksync.io/docs/tools/testing/dockerized-testing.html) by removing the geth service altogether in docker-compose.yml (as I have my own L1 node) and pointing ETH_CLIENT_WEB3_URL to my local L1 node instance. When I run “docker compose up” the docker service is not able to connect to my local L1 node as specified by ETH_CLIENT_WEB3_URL. For example, all L1 contract deployments result in “Error: could not detect network (event="noNetwork", code=NETWORK_ERROR, version=providers/5.7.2)”. I haven’t yet investigated deeper into the error, as the execution flow is buried in external Docker images.

Before I commit further to this path, I’d like to ask this community if anyone has tried something like this, whether there might be a better approach, or whether I may have missed relevant documentation.

Another way to ask is: How does one instantiate and deploy an zkSync Era rollup to a brand new EVM-compatible L1 chain (not geth client)?

### What is the specific issue or error you're encountering?

I experimented with the [Dockerized setup](https://era.zksync.io/docs/tools/testing/dockerized-testing.html) by removing the geth service altogether in docker-compose.yml (as I have my own L1 node) and pointing `ETH_CLIENT_WEB3_URL` to my local L1 node instance. When I run “`docker compose up`” the docker service is not able to connect to my local L1 node as specified by `ETH_CLIENT_WEB3_URL`. For example, all L1 contract deployments result in “`Error: could not detect network (event="noNetwork", code=NETWORK_ERROR, version=providers/5.7.2)`”

### Can you share the error messages or logs you're receiving, if any?

Here is a section of the logs from `docker compose up`:

```
local-setup-zksync-1	| > Deploying L1 verifier
local-setup-zksync-1	| $ ts-node scripts/deploy.ts --only-verifier
local-setup-zksync-1	| Using deployer wallet: 0x52312AD6f01657413b2eaE9287f6B9ADaD93D5FE
local-setup-zksync-1	| Using governor address: 0x52312AD6f01657413b2eaE9287f6B9ADaD93D5FE
local-setup-zksync-1	| Error: Error: could not detect network (event="noNetwork", code=NETWORK_ERROR, version=providers/5.7.2)
local-setup-zksync-1	| 	at Logger.makeError (/contracts/ethereum/node_modules/@ethersproject/logger/src.ts/index.ts:269:28)
local-setup-zksync-1	| 	at Logger.throwError (/contracts/ethereum/node_modules/@ethersproject/logger/src.ts/index.ts:281:20)
local-setup-zksync-1	| 	at JsonRpcProvider.<anonymous> (/contracts/ethereum/node_modules/@ethersproject/providers/src.ts/json-rpc-provider.ts:483:23)
local-setup-zksync-1	| 	at step (/contracts/ethereum/node_modules/@ethersproject/providers/lib/json-rpc-provider.js:48:23)
local-setup-zksync-1	| 	at Object.throw (/contracts/ethereum/node_modules/@ethersproject/providers/lib/json-rpc-provider.js:29:53)
local-setup-zksync-1	| 	at rejected (/contracts/ethereum/node_modules/@ethersproject/providers/lib/json-rpc-provider.js:21:65)
local-setup-zksync-1	| 	at processTicksAndRejections (node:internal/process/task_queues:95:5) {
local-setup-zksync-1	|   reason: 'could not detect network',
local-setup-zksync-1	|   code: 'NETWORK_ERROR',
local-setup-zksync-1	|   event: 'noNetwork'
local-setup-zksync-1	| }
error Command failed with exit code 1.
```

### Have you made any recent changes to the contract before encountering this issue?

No

### Are there any external libraries or contracts that your contract interacts with?

Default

### Can you provide the relevant portions of your contract code where the issue is occurring?

Following instructions here: [Dockerized setup](https://era.zksync.io/docs/tools/testing/dockerized-testing.html)

### Have you tried to isolate the problem, and if so, what were the results?

Please see the main description.

### What steps have you already taken to try to resolve the issue?

Basic checks, ensuring that my L1 node is running and that `ETH_CLIENT_WEB3_URL` is configured correctly.

### Repo Link (Optional)

https://github.com/matter-labs/local-setup.git

### Additional Details

I’d like to ask this community if anyone has tried something like this, whether there might be a better approach, or whether I may have missed relevant documentation.

Another way to ask is: How does one instantiate and deploy an zkSync Era rollup to a brand new EVM-compatible L1 chain (not geth client)?

Thanks.

**Comment by dutterbutter** [comment, marked as answer]
I dont think I've seen anyone do this yet or at the very least post about it. 

The dockerized setup is made up of a few different services that all expect `geth` which is why you are probably having issues. You would have to remove all the dependencies on `geth` from all services and replace with your own L1 (e.g. `docker-compose-runner.yml`, `matterlabs/zk-environment`,  etc). 

Even then I am not sure if it will work depending on your L1 setup. Will leave this up to see if someone from the community has made similar efforts.

**Reply by AlexRamRam** [reply]
Thank you for your reply, Dustin.

As a related question, do you know of any specific examples that illustrate how these services are tied specifically to a Geth implementation? Naively, one might expect that any Ethereum JSON-RPC compliant node would suffice.

**Reply by dutterbutter** [reply]
Unfortunately there are no specific examples, and although you are correct that you would need Ethereum JSON-RPC compliant node, I also believe there would be a need to make changes to the proving system which would not be trivial to make. I will investigate to see if I can get more information here but I do think this would be a larger effort then I initially considered.

**Comment by amyliasnow** [comment]
It seems like you're encountering a network connection issue when trying to connect the zkSync Era Dockerized setup to your local L1 node. Here are a few steps you can take to troubleshoot:

Double-check the configuration of ETH_CLIENT_WEB3_URL in your Docker environment. Ensure that it points to the correct URL of your local L1 node and that the node is accessible from within the Docker container.

Verify that your local L1 node is running and accessible. You can test this by trying to connect to it using a tool like curl or wget from within the Docker container.

Check for any firewall or network configuration issues that might be blocking the connection between the Docker container and your local L1 node.

Review the Docker logs and error messages for more detailed information about the network error. It might provide clues about what's causing the connection failure.

Consider reaching out to the zkSync community or checking their documentation for any specific instructions or best practices regarding connecting the zkSync Era Dockerized setup to a custom L1 node.

By following these steps and potentially seeking assistance from the zkSync community, you should be able to diagnose and resolve the network connection issue.

**Comment by ryangrullon** [comment]
It seems like the local L1 node is not properly configured in the Dockerized setup, leading to the "noNetwork" error. Double-check the configuration of ETH_CLIENT_WEB3_URL and ensure that the local L1 node is accessible from within the Docker container. Additionally, review the Dockerfile and docker-compose.yml to ensure they are correctly configured to interact with the local L1 node.

**Comment by reginaldolivas** [comment]
It seems there's an issue connecting the Dockerized zkSync setup to a custom EVM-compatible L1 chain. You may need to ensure that the configuration for connecting to the local L1 node is correct and that the network settings are properly configured in the Docker environment. Additionally, consulting with the zkSync community or documentation for specific guidance on deploying to custom EVM-compatible chains could be helpful.
